include/cuco/detail/dynamic_map_kernels.cuh

```diff
@@ -487,6 +487,85 @@ CUCO_KERNEL void find(InputIt first,
   }
 }
 
+/**
+ * @brief Retrieves all of the keys and their associated values.
+ *
+ * The order in which keys are returned is implementation defined and not guaranteed to be
+ * consistent between subsequent calls to `retrieve_all`.
+ *
+ * Behavior is undefined if the range beginning at `keys_out` or `values_out` is less than
+ * `get_size()`
+ *
+ * @tparam block_size The number of threads in the thread block
+ * @tparam KeyOutputIt Device accessible output iterator whose `value_type` is
+ * convertible to the map's `key_type`
+ * @tparam ValueOutputIt Device accessible output iterator whose `value_type` is
+ * convertible to the map's `mapped_type`
+ * @tparam viewT Type of `static_map` device view
+ * @tparam PrefixT Type of the prefix sum array of submaps
+ *
+ * @param keys_out Beginning output iterator for keys
+ * @param values_out Beginning output iterator for values
+ * @param submap_views Array of `static_map::device_view` objects used to
+ * perform `retrieve_all` operations on each underlying `static_map`
+ * @param num_submaps The number of submaps in the map
+ * @param capacity The total number of slots of all submaps
+ * @param d_num_out Pointer to the device memory location where the number of keys/vals retrieved
+ * are stored
+ * @param cap_prefix_sum Array of prefix sums of the number of slots in each submap
+ * @return Pair of iterators indicating the last elements in the output
+ */
+template <uint32_t block_size,
+          typename KeyOutputIt,
+          typename ValueOutputIt,
+          typename viewT,
+          typename PrefixT>
+CUCO_KERNEL void retrieve_all(KeyOutputIt keys_out,
+                              ValueOutputIt values_out,
+                              viewT* submap_views,
+                              uint32_t num_submaps,
+                              uint64_t capacity,
+                              unsigned long long* d_num_out,
+                              PrefixT* cap_prefix_sum)
+{
+  using BlockScan = cub::BlockScan<unsigned int, block_size>;
+
+  __shared__ typename BlockScan::TempStorage scan_temp_storage;
+  __shared__ unsigned int block_base;
+
+  auto tid                       = blockDim.x * blockIdx.x + threadIdx.x;
+  auto const empty_key_sentinel  = submap_views[0].get_empty_key_sentinel();
+  auto const erased_key_sentinel = submap_views[0].get_erased_key_sentinel();
+
+  while (tid < capacity) {
```

Comment:
```suggestion
  while ((tid - threadIdx.x) < capacity) {
```
We need to ensure that the entire thread block follows the same execution path; otherwise, the `__syncthreads()` at line 558 could fail if some threads don't enter the while loop. The handling around `tid` also needs to be updated accordingly to reflect this change.